src/libopensc/card-edo.c

```diff
@@ -0,0 +1,332 @@
+/*
+ * Copyright (C) 2020 Piotr Majkrzak
+ *
+ * This file is part of OpenSC.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#if defined(ENABLE_SM) && defined(ENABLE_OPENPACE)
+
+#include "libopensc/internal.h"
+#include "libopensc/opensc.h"
+#include "libopensc/pace.h"
+#include "libopensc/sm.h"
+#include "libopensc/asn1.h"
+#include "sm/sm-eac.h"
+#include <string.h>
+#include <stdlib.h>
+
+
+static struct sc_card_operations edo_ops;
+
+
+static struct sc_card_driver edo_drv = {
+	"Polish eID card (e-dowód, eDO)",
+	"edo",
+	&edo_ops,
+	NULL, 0, NULL
+};
+
+
+static const struct sc_atr_table edo_atrs[] = {
+	{ "3b:84:80:01:47:43:50:43:12", NULL, NULL, SC_CARD_TYPE_EDO, 0, NULL },
+	{ NULL, NULL, NULL, 0, 0, NULL }
+};
+
+
+static struct {
+	int len;
+	struct sc_object_id oid;
+} edo_curves[] = {
+	// secp384r1
+	{384, {{1, 3, 132, 0, 34, -1}}}
+};
+
+
+static void edo_eac_init() {
+	extern void EAC_init(void);
+	static int initialized = 0;
+	if (!initialized) {
+		EAC_init();
+		initialized = 1;
+	}
+}
+
+
+static int edo_match_card(sc_card_t* card) {
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+	if (_sc_match_atr(card, edo_atrs, &card->type) >= 0) {
+		sc_log(card->ctx, "ATR recognized as Polish eID card\n");
+		LOG_FUNC_RETURN(card->ctx, 1);
+	}
+	LOG_FUNC_RETURN(card->ctx, 0);
+}
+
+
+static int edo_get_can(sc_card_t* card, struct establish_pace_channel_input* pace_input) {
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+	const char* can;
+
+	if (!(can = getenv("EDO_CAN"))) {
+		for (size_t i = 0; card->ctx->conf_blocks[i]; ++i) {
+			scconf_block** blocks = scconf_find_blocks(card->ctx->conf, card->ctx->conf_blocks[i], "card_driver", "edo");
+			if (!blocks)
+				continue;
+			for (size_t j = 0; blocks[j]; ++j)
+				if ((can = scconf_get_str(blocks[j], "can", NULL)))
```

Comment:
please add this to the documentation of `opensc.conf` (https://github.com/OpenSC/OpenSC/blob/master/doc/files/opensc.conf.5.xml.in) and maybe in the example file https://github.com/OpenSC/OpenSC/blob/master/etc/opensc.conf.example.in